src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -1234,10 +1236,121 @@ TargetListOnPartitionColumn(Query *query, List *targetEntryList)
 		}
 	}
 
+	/*
+	 * We could stil behave as if the target list is on partition column if
+	 * all range table entries are reference tables and there is no column
+	 * references to any outer queries.
+	 */
+	if (!targetListOnPartitionColumn)
+	{
+		if (RangeTableListContainsOnlyReferenceTables(query->rtable) &&
+			!TargetListContainsNotSupportedEntries(targetEntryList))
+		{
+			targetListOnPartitionColumn = true;
+		}
+	}
+
 	return targetListOnPartitionColumn;
 }
 
 
+/*
+ * TargetListContainsNotSupportedEntries returns true if non of the
+ * elements in the target entry list belong to an outer query (for
+ * example the query is a sublink and references to another query
+ * in the from list).
+ *
+ * The function also returns true if any of the  target entries is not
+ * a column itself. This might be too restrictive, but, given that we're
+ * handling a very specific type of queries, that seems acceptable for now.
+ */
+static bool
+TargetListContainsNotSupportedEntries(List *targetEntryList)
```

Comment:
I did not like this name. It would be better if we name this function as what it actually does. Implementation does not look like it care about being supported or not supported.

It is more like
```c
bool AllTargetExpressionsAreColumnReferences(List *targetEntryList)
```

Reply:
Makes sense, updating the name.